### PR TITLE
ocamlPackages.happy-eyeballs: 0.1.3 -> 0.3.0

### PR DESCRIPTION
**Pull request by ulrikstrid**
###### Description of changes

Update happy-eyeballs and add lwt and mirage packages

Needs https://github.com/NixOS/nixpkgs/pull/192256

Fixes #184393

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/happy-eyeballs/default.nix [hunk 1]
```diff
@@ -4,16 +4,18 @@
...
   src = fetchurl {
     url = "https://github.com/roburio/happy-eyeballs/releases/download/v${version}/happy-eyeballs-${version}.tbz";
-    sha256 = "sha256:0ns1bxcmx0rkq4am6vl2aargdzkfhria8sfmgnh8dgzvvj93cc1c";
+    sha256 = "17mnid1gvq1ml1zmqzn0m6jmrqw4kqdrjqrdsrphl5kxxyhs03m6";
   };
 
-  propagatedBuildInputs = [
+  strictDeps = true;
+
+  buildInputs = [
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/default.nix:18]:
```suggestion
  propagatedBuildInputs = [
```
"domain-name duration fmt ipaddr logs" should be propagated.

File: pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  buildInputs = [
+    happy-eyeballs
+    cmdliner
+    duration
+    dns-client
+    domain-name
+    ipaddr
+    fmt
+    logs
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix:23]:
```suggestion
    dns-client
    logs
```

"domain-name duration fmt ipaddr logs" is being propagated by "happy-eyeballs".

File: pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{ lib, buildDunePackage
+, happy-eyeballs, cmdliner, duration, dns-client
+, domain-name, ipaddr, fmt, logs, lwt, mtime
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix:3]:
```suggestion
, happy-eyeballs, cmdliner, dns-client, lwt, mtime
```

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  strictDeps = true;
+
+  propagatedBuildInputs = [
+    happy-eyeballs
+    duration
+    dns-client
+    domain-name
+    ipaddr
+    fmt
+    logs
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:32]:
```suggestion
    dns-client
```

`duration domain-name ipaddr fmt logs` is to be propagated by `happy-eyeballs`

**Comment by vbgl** [comment]
@GrahamcOfBorg build happy-eyeballs-mirage

**Comment by superherointj** [comment]
For me, it is failing to build:
* ocamlPackages.happy-eyeballs-lwt
  image

* ocamlPackages.happy-eyeballs-mirage
  image

**Comment by ulrikstrid** [comment]
This should be good to go now @vbgl @superherointj 

After that I can rebase #180726 again

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 180724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jackline</li>
  </ul>
</details>

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 180724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jackline</li>
  </ul>
</details>

File: pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix [hunk 5]
```diff
@@ -4,23 +4,29 @@
...
-  propagatedBuildInputs = [
+  buildInputs = [
     dns-client
     happy-eyeballs
+    cmdliner
     logs
     lwt
+    mtime
   ];
+  doCheck = false;
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix:27]:
```suggestion
```
I believe `doCheck = false;` is default.

File: pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix [hunk 6]
```diff
@@ -4,23 +4,29 @@
...
+
   inherit (happy-eyeballs) src version;
 
-  buildInputs = [ cmdliner ];
+  minimalOCamlVersion = "4.08";
+
+  strictDeps = true;
 
-  propagatedBuildInputs = [
+  buildInputs = [
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix:19]: image

These should be at `propagatedBuildInputs`:
* dns-client
* happy-eyeballs
* logs
* lwt

File: pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix [hunk 7]
```diff
@@ -4,23 +4,29 @@
...
     dns-client
     happy-eyeballs
+    cmdliner
     logs
     lwt
+    mtime
   ];
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/lwt.nix:26]:
```suggestion
  buildInputs = [
    cmdliner
    mtime
  ];
  
  propagatedBuildInputs = [
    dns-client
    happy-eyeballs
    logs
    lwt
  ];

Source of suggestion:
  ![image](https://user-images.githubusercontent.com/5861043/191563564-765b6b2f-64f3-4319-90f1-0e0dc1c9c97d.png)

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix [hunk 8]
```diff
@@ -1,25 +1,39 @@
...
     dns-client
     happy-eyeballs
-    logs
     lwt
+    mirage-clock
+    mirage-random
+    mirage-time
     tcpip
   ];
+  doCheck = false;
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:34]:
```suggestion
```
`doCheck = false;` is default.

Comment (superherointj) [reply pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:34]: Unless someone plans on reverting that in the future.
Would make sense default to be true (in the future).
@vbgl Is there plans for that?

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix [hunk 9]
```diff
@@ -1,25 +1,39 @@
...
   propagatedBuildInputs = [
     dns-client
     happy-eyeballs
-    logs
     lwt
+    mirage-clock
+    mirage-random
+    mirage-time
     tcpip
   ];
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:33]: image

`logs` is missing from propagatedBuildInputs

mirage-clock, mirage-random, mirage-time is not propagated.

Comment (superherointj) [reply pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:33]:
```
buildInputs = [
  mirage-clock
  mirage-random
  mirage-time
];
```

Opam package has:
* duration
* domain-name
* ipaddr
* fmt (?)

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix [hunk 10]
```diff
@@ -1,25 +1,39 @@
...
+  doCheck = false;
 
-  meta = happy-eyeballs.meta // {
+  meta =  happy-eyeballs.meta // {
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:36]:
```suggestion
  meta = happy-eyeballs.meta // {
```
just a blank space

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix [hunk 11]
```diff
@@ -1,25 +1,39 @@
 { buildDunePackage
 , happy-eyeballs
+, duration
```
Comment (superherointj) [inline pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:3]:
```suggestion
, duration
```
Where is duration used here?

Comment (superherointj) [reply pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:3]: `logs` not used too.
domain-name, fmt, ipaddr. maybe some more?

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix (hunk 9 above)
Comment (superherointj) [reply pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix:33]: `lwt` is already propagated by dns-client.

**Comment by superherointj** [comment]
At `happy-eyeballs`, the package `ipaddr` already propagates `domain-name`.

File: pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix (hunk 8 above)
Comment (ulrikstrid) [reply pkgs/development/ocaml-modules/happy-eyeballs/mirage.nix]: Looks like we can run tests now so I enabled them

**Comment by ulrikstrid** [comment]
I went through the modules and compared to opam so now it should be fine @superherointj

**Comment by superherointj** [comment]
> I went through the modules and compared to opam so now it should be fine @superherointj

Not opam. But what is actually used. In the case, you should be checking the META file and dependencies.

**Comment by vbgl** [comment]
> Do you agree dependencies that are already propagated by some other package dependency should not be included?

No.

**Pull request merged**